pages/_document.js

```diff
@@ -1,27 +1,54 @@
 import Document, { Html, Head, Main, NextScript } from 'next/document'
+import { generateNonce } from '../lib/Utils.js'
 import Script from 'next/script'
 
 class MyDocument extends Document {
   static async getInitialProps(ctx) {
+    const nonce = generateNonce()
     const initialProps = await Document.getInitialProps(ctx)
-    return { ...initialProps }
+    //set csp
+    let csp = ``
+    csp += `base-uri 'self';`
+    csp += `form-action 'self';`
+    csp += `default-src 'self';`
+    csp += `script-src 'self' 'nonce-${nonce}' 'unsafe-eval';`
+    csp += `style-src 'self' https://fonts.googleapis.com 'unsafe-inline';`
+    csp += `img-src 'self' data: blob:;`
+    csp += `font-src 'self' https://fonts.gstatic.com data:;`
+    csp += `frame-ancestors 'self';`
+
+    if (ctx.res) {
+      ctx.res.setHeader('Content-Security-Policy', csp)
+    }
+
+    return { ...initialProps, nonce }
```

Comment:
If possible it is better to set headers in one spot. CSP might be getting overwritten in the next.config.js file

  {
    key: 'Content-Security-Policy',
    value:
      "default-src 'self'; base-uri 'self'; frame-ancestors 'self'; form-action 'self'; object-src 'none'; script-src-elem 'self'; script-src 'self' 'unsafe-eval'; connect-src 'self'; style-src 'self' 'unsafe-inline' https://fonts.googleapis.com; font-src 'self' https://fonts.gstatic.com; img-src 'self' data: https:",
  },

Reply:
Because some of the headers have a dynamic nonce or hash I believe that part needs to remain still in _document. I've moved our static CSP into the config as well and just appended script-src in _document but there are other ways to do this as well.